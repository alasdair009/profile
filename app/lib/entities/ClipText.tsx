"use client";

import styled from "styled-components";
import {rem} from "polished";

export const ClipText = styled.span`
  align-items: center;
  //background: black;
  display: flex;
  font-size: ${rem(56)};
  height: 100%;
  justify-content: center;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 100%;
  -webkit-text-fill-color: transparent;
  text-fill-color: transparent;
  -webkit-background-clip: text;
  background-clip: text;
  background: black;
  color: white;
`;