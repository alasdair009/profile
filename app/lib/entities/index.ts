"use client";

import styled from "styled-components";

export * from "@/entities";
export { Splash } from "./Splash";

export const FireWrapper = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  position: absolute;
  width: 100%;
`;
export const FirePadding = styled.div<{ $backgroundColor: string }>`
  background: ${({ $backgroundColor }) => $backgroundColor};
  flex: 1;
`;
