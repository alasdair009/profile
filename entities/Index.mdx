import { Meta } from "@storybook/blocks";
import Image from "next/image";
import amLogo from "./assets/am.svg";
import { rem } from "polished";

<Meta title="Alasdair Macrae Design" />

# Alasdair Macrae Design

<Image
  src={amLogo}
  alt="Green and white AM letters"
  width={0}
  height={0}
  style={{ maxWidth: rem(240), width: "100%", height: "auto" }}
/>

I am a keen supported of Storybook js using it across the majority of projects I build from websites, web-apps and even desktop applications.
