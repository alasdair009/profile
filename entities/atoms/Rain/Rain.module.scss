@use "../../../app/styles/dimensions";

.rainCanvasWrapper {
  height: 100%;
  min-height: dimensions.getSize(s128);
  position: absolute;
  width: 100%;
}

.rainCanvas {
  height: 100%;
  position: absolute;
  width: 100%;
}
