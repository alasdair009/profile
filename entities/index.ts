export { borderRadii } from "./design-tokens/effects";
export { colors } from "./design-tokens/colors";
export { fonts } from "./design-tokens/typography";
export { sizes } from "./design-tokens/dimensions";

export { Button } from "./atoms/Button";
export { Fire } from "./atoms/Fire";
export { Heading } from "./atoms/Heading";
export { Link } from "./atoms/Link";
export { Orb } from "./atoms/Orb";
export { Paragraph } from "./atoms/Paragraph";
export { Spacer } from "./atoms/Spacer";
export { TextMask } from "./atoms/TextMask";

export { Container } from "./organisms/Container";
export { ContentPlate } from "./organisms/ContentPlate";
export { Header } from "./organisms/Header";
export { Splash } from "./organisms/Splash";
