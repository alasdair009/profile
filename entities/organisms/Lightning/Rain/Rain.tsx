"use client";
import { HTMLAttributes, useEffect, useRef } from "react";
import { RainCanvasWrapper as Root, RainCanvas } from "@/entities/organisms/Lightning/styles";
import { lighten, rgba } from "polished";

type RainProps = {
  rainDrops?: number;
  rainColor?: string;
  speedRainTrough?: number;
} & HTMLAttributes<HTMLDivElement>;

type RainDrop = {
  x: number;
  y: number;
  l: number;
  xs: number;
  ys: number;
};

type RainTrough = {
  x: number;
  y: number;
  length: number;
  opacity: number;
  xs: number;
  ys: number;
};

function random(min: number, max: number) {
  return Math.random() * (max - min + 1) + min;
}

/**
 * A canvas playing a rain animation
 */
export function Rain({ rainColor = "rgb(174,194,224)", rainDrops = 200, speedRainTrough = 25, ...rest }: RainProps) {
  const rainCanvasRef = useRef<HTMLCanvasElement | null>(null);
  const rain: RainDrop[] = [];
  const rainTrough: RainTrough[] = [];
  const troughColor = lighten(0.02, rainColor);

  const drawRain = (ctx: CanvasRenderingContext2D, i: number) => {
    ctx.beginPath();
    ctx.moveTo(rain[i].x, rain[i].y);
    ctx.lineTo(rain[i].x + rain[i].l * rain[i].xs, rain[i].y + rain[i].l * rain[i].ys);
    ctx.strokeStyle = rgba(rainColor, 0.5);
    ctx.lineWidth = 1;
    ctx.lineCap = "round";
    ctx.stroke();
  };

  const drawRainTrough = (ctx: CanvasRenderingContext2D, i: number) => {
    ctx.beginPath();
    const grd = ctx.createLinearGradient(0, rainTrough[i].y, 0, rainTrough[i].y + rainTrough[i].length);
    grd.addColorStop(0, rgba(troughColor, 0));
    grd.addColorStop(1, rgba(troughColor, rainTrough[i].opacity));

    ctx.fillStyle = grd;
    ctx.fillRect(rainTrough[i].x, rainTrough[i].y, 1, rainTrough[i].length);
    ctx.fill();
  };

  const clear = (ctx: CanvasRenderingContext2D, w: number, h: number) => {
    ctx.clearRect(0, 0, w, h);
  };

  const createRain = (w: number, h: number) => {
    for (let i = 0; i < rainDrops; i++) {
      rain[i] = {
        x: Math.random() * w,
        y: Math.random() * h,
        l: Math.random(),
        xs: -4 + Math.random() * 4 + 2,
        ys: Math.random() * 10 + 10,
      };
    }
  };

  const createRainTrough = (w: number, h: number) => {
    for (let i = 0; i < rainDrops; i++) {
      rainTrough[i] = {
        x: random(0, w),
        y: random(0, h),
        length: Math.floor(random(1, 830)),
        opacity: Math.random() * 0.2,
        xs: random(-2, 2),
        ys: random(10, 20),
      };
    }
  };

  const animate = (ctx: CanvasRenderingContext2D, w: number, h: number) => {
    clear(ctx, w, h);

    for (let i = 0; i < rainDrops; i++) {
      rain[i].x += rain[i].xs;
      rain[i].y += rain[i].ys;
      if (rain[i].x > w || rain[i].y > h) {
        rain[i].x = Math.random() * w;
        rain[i].y = -20;
      }
      drawRain(ctx, i);

      if (rainTrough[i].y >= h) {
        rainTrough[i].y = h - rainTrough[i].y - rainTrough[i].length * 5;
      } else {
        rainTrough[i].y += speedRainTrough;
      }
      drawRainTrough(ctx, i);
    }
  };

  useEffect(() => {
    const rainCanvas = rainCanvasRef.current;
    if (!rainCanvas) {
      return;
    }

    const rainContext = rainCanvas.getContext("2d");
    if (!rainContext) {
      return;
    }

    let w = rainCanvas.width;
    let h = rainCanvas.height;
    window.addEventListener("resize", function () {
      w = rainCanvas.width;
      h = rainCanvas.height;
      createRainTrough(w, h);
    });

    createRain(w, h);
    createRainTrough(w, h);

    const animloop = () => {
      animate(rainContext, w, h);
      requestAnimationFrame(animloop);
    };
    animloop();
  }, [drawRain]);

  return (
    <Root {...rest}>
      <RainCanvas ref={rainCanvasRef} />
    </Root>
  );
}
