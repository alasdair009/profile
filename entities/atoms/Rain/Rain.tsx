"use client";
import { HTMLAttributes, useEffect, useRef } from "react";
import { lighten, rgba } from "polished";
import { prefersReducedMotion } from "@/entities";
import { LCG, lcgNextRand, makeLCG } from "@/lib/random";
import styles from "./Rain.module.scss";

type RainProps = {
  rainDrops?: number;
  rainColor?: string;
  speedRainTrough?: number;
} & HTMLAttributes<HTMLDivElement>;

type RainDrop = {
  x: number;
  y: number;
  l: number;
  xs: number;
  ys: number;
};

type RainTrough = {
  x: number;
  y: number;
  length: number;
  opacity: number;
  xs: number;
  ys: number;
};

function random(lcg: LCG, min: number, max: number) {
  return lcgNextRand(lcg) * (max - min + 1) + min;
}

/**
 * A canvas playing a rain animation
 */
export function Rain({
  rainColor = "rgb(174,194,224)",
  rainDrops = 200,
  speedRainTrough = 25,
  ...rest
}: RainProps) {
  const rainLCG = makeLCG();
  const rainCanvasRef = useRef<HTMLCanvasElement | null>(null);
  const rain: RainDrop[] = [];
  const rainTrough: RainTrough[] = [];
  const troughColor = lighten(0.02, rainColor);

  const drawRain = (ctx: CanvasRenderingContext2D, i: number) => {
    ctx.beginPath();
    ctx.moveTo(rain[i].x, rain[i].y);
    ctx.lineTo(
      rain[i].x + rain[i].l * rain[i].xs,
      rain[i].y + rain[i].l * rain[i].ys
    );
    ctx.strokeStyle = rgba(rainColor, 0.5);
    ctx.lineWidth = 1;
    ctx.lineCap = "round";
    ctx.stroke();
  };

  const drawRainTrough = (ctx: CanvasRenderingContext2D, i: number) => {
    ctx.beginPath();
    const grd = ctx.createLinearGradient(
      0,
      rainTrough[i].y,
      0,
      rainTrough[i].y + rainTrough[i].length
    );
    grd.addColorStop(0, rgba(troughColor, 0));
    grd.addColorStop(1, rgba(troughColor, rainTrough[i].opacity));

    ctx.fillStyle = grd;
    ctx.fillRect(rainTrough[i].x, rainTrough[i].y, 1, rainTrough[i].length);
    ctx.fill();
  };

  const clear = (ctx: CanvasRenderingContext2D, w: number, h: number) => {
    ctx.clearRect(0, 0, w, h);
  };

  const createRain = (w: number, h: number) => {
    for (let i = 0; i < rainDrops; i++) {
      rain[i] = {
        x: lcgNextRand(rainLCG) * w,
        y: lcgNextRand(rainLCG) * h,
        l: lcgNextRand(rainLCG),
        xs: -4 + lcgNextRand(rainLCG) * 4 + 2,
        ys: lcgNextRand(rainLCG) * 10 + 10,
      };
    }
  };

  const createRainTrough = (w: number, h: number) => {
    for (let i = 0; i < rainDrops; i++) {
      rainTrough[i] = {
        x: random(rainLCG, 0, w),
        y: random(rainLCG, 0, h),
        length: Math.floor(random(rainLCG, 1, 830)),
        opacity: lcgNextRand(rainLCG) * 0.2,
        xs: random(rainLCG, -2, 2),
        ys: random(rainLCG, 10, 20),
      };
    }
  };

  const animate = (ctx: CanvasRenderingContext2D, w: number, h: number) => {
    clear(ctx, w, h);

    for (let i = 0; i < rainDrops; i++) {
      if (!prefersReducedMotion) {
        rain[i].x += rain[i].xs;
        rain[i].y += rain[i].ys;
        if (rain[i].x > w || rain[i].y > h) {
          rain[i].x = lcgNextRand(rainLCG) * w;
          rain[i].y = -20;
        }
      }
      drawRain(ctx, i);

      if (!prefersReducedMotion) {
        if (rainTrough[i].y >= h) {
          rainTrough[i].y = h - rainTrough[i].y - rainTrough[i].length * 5;
        } else {
          rainTrough[i].y += speedRainTrough;
        }
      }
      drawRainTrough(ctx, i);
    }
  };

  useEffect(() => {
    const rainCanvas = rainCanvasRef.current;
    if (!rainCanvas) {
      return;
    }

    const rainContext = rainCanvas.getContext("2d");
    if (!rainContext) {
      return;
    }

    let w = rainCanvas.width;
    let h = rainCanvas.height;
    window.addEventListener("resize", function () {
      w = rainCanvas.width;
      h = rainCanvas.height;
      createRainTrough(w, h);
    });

    createRain(w, h);
    createRainTrough(w, h);

    const animloop = () => {
      animate(rainContext, w, h);
      requestAnimationFrame(animloop);
    };
    animloop();
  }, [drawRain]);

  return (
    <div className={styles.rainCanvasWrapper} data-testid={Rain.name} {...rest}>
      <canvas
        className={styles.rainCanvas}
        ref={rainCanvasRef}
        data-chromatic="ignore"
      />
    </div>
  );
}
