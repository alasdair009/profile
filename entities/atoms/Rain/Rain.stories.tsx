import { Rain } from "./Rain";
import { Meta, StoryObj } from "@storybook/nextjs";
import { colors } from "@/entities";
import { expect, within } from "storybook/test";

const meta: Meta<typeof Rain> = {
  component: Rain,
  args: {
    rainDrops: 200,
    rainColor: "rgb(174,194,224)",
    speedRainTrough: 25,
  },
  argTypes: {
    rainDrops: {
      control: { type: "range", min: 10, max: 500, step: 1 },
    },
    rainColor: {
      control: {
        type: "color",
        presetColors: ["rgb(174,194,224)", colors.greenGrass],
      },
    },
    speedRainTrough: {
      control: { type: "range", min: 1, max: 25, step: 1 },
    },
  },
};
export default meta;

export const Default: StoryObj<typeof Rain> = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const rainElement = canvas.getByTestId(Rain.name);

    await expect(rainElement).toBeInTheDocument();
  },
};
