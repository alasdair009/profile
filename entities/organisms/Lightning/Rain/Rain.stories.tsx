import { Rain } from "./Rain";
import { Meta, StoryObj } from "@storybook/react";
import { colors } from "@/entities";

const meta: Meta<typeof Rain> = {
  component: Rain,
  parameters: {
    docs: {
      inlineStories: false,
      iframeHeight: 400,
    },
  },
  args: {
    rainDrops: 200,
    rainColor: "rgb(174,194,224)",
  },
  argTypes: {
    rainDrops: {
      control: { type: "range", min: 10, max: 500, step: 1 },
    },
    rainColor: {
      control: {
        type: "color",
        presetColors: ["rgb(174,194,224)", colors.greenGrass],
      },
    },
  },
};
export default meta;

export const Default: StoryObj<typeof Rain> = {};
