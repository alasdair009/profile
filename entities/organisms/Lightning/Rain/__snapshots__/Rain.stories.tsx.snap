// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organisms/Lightning/Rain Default smoke-test 1`] = `
<div class="sc-iAlELC gRILcD">
  <canvas class="sc-oeqTF hjFjiZ">
  </canvas>
</div>
`;
